### ISSUE TITLE
ansible 2.0.0.2 - ValueError: file descriptor out of range in select()

### ISSUE THREAD

**Issue by bbyhuy** [ansible/ansible#14143]
**Issue Type:**
Bug Report

**Ansible Version:**
2.0.0.2

**Ansible Configuration:**
- Enabled Logging
- Disabled host-key checking

**Environment:**
- Ubuntu 14.04 KVM

**Summary:**
Attempting to run a playbook across 220 VMs running ubuntu at 220 forks. Started getting and error complaining about descriptor limit. So I raised the ulimit (via /etc/security/limits.conf) and confirmed via `ulimit -n` that it is at 100000. Still getting error, raised to 200000, no go. 

**Steps to Reproduce:**
Spawn 220 virtual machines.
Run the following

```
ansible all -i 220VMhosts -b -f 220 -m shell -a 'echo hello'
```

Will see error. 

**Expected Results:**
Successful hellos across all 220 VMs. 

**Actual Results:**

```
2016-01-26 16:41:22,990 p=8627 u=root |  ^[[0;31mAn exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/process/worker.py", line 114, in run
    self._shared_loader_obj,
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 91, in run
    item_results = self._run_loop(items)
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 241, in _run_loop
    res = self._execute(variables=task_vars)
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 401, in _execute
    result = self._handler.run(task_vars=variables)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/action/normal.py", line 33, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars))
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/action/__init__.py", line 395, in _execute_module
    tmp = self._make_tmp_path()
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/action/__init__.py", line 210, in _make_tmp_path
    result = self._low_level_execute_command(cmd, sudoable=False)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/action/__init__.py", line 510, in _low_level_execute_command
    rc, stdout, stderr = self._connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/connection/ssh.py", line 600, in exec_command
    return_tuple = self._exec_command(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/connection/ssh.py", line 577, in _exec_command
    (returncode, stdout, stderr) = self._run(cmd, in_data, sudoable=sudoable)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/connection/ssh.py", line 417, in _run
    rfd, wfd, efd = select.select(rpipes, [], [], timeout)
ValueError: filedescriptor out of range in select()
```

**Comment by chrrrles** [comment]
Can you also run `ulimit -n` on the machine running ansible-playbook?

Possibly related to:  https://github.com/ansible/ansible/issues/10157

**Comment by bbyhuy** [comment]
Updated. Sorry about that. Let me know if anything else is needed.

**Comment by kai11** [comment]
Same problem (with fix) in supervisor https://github.com/Supervisor/supervisor/pull/129

**Comment by kai11** [comment]
@bbyhuy I've made very quick and dirty fix for this problem.
Replace ssh.py of your Ansible installation with my file and test.
https://github.com/kai11/ansible/blob/fix/select_fd_out_of_range_wip/lib/ansible/plugins/connection/ssh.py

**Comment by bcoca** [comment]
see if this PR fixes it for you #14458

**Comment by jgoldschrafe** [comment]
@bcoca, as someone else with this problem: tested PR, still receiving error.

**Comment by jgoldschrafe** [comment]
The quick-and-dirty poller implementation from @kai11 works fine, though.

**Comment by kai11** [comment]
Some background info on select vs poll
http://stackoverflow.com/questions/4039832/select-vs-poll-vs-epoll
http://daniel.haxx.se/docs/poll-vs-select.html

Quote from last link:
**Portability**
select - has existed for a great while and exists almost everywhere. **A problem with many file descriptors is that you cannot know if you will overflow the the bitmask as you can't check the file descriptor againt FD_SETSIZE in a portable manner.**
poll - **Not existing on older unixes nor in Windows before Vista. Broken implementation in Mac OS X at least up to 10.4 and earlier.** Up to 10.3 you could check for a poll() that works with all arguments zeroed. The 10.4 bug is more obscure an I don't know of any way to detect it without checking for OS.

---

Not sure if MacOs bug note applies to Python.
In Ubuntu 14.04 Python binary have FD_SETSIZE set to 1024, very low.

I see two options:
1) Use kqueue (for MacOS), poll() (for Linux) and fallback to select (windows). Ignore epoll, etc... - they will decrease compatibility.
2) Add pyev dependency and have 100% compatibility with everything. Although pypi library is from 2013.

Of course, with fixes from #14458

@bcoca Thoughts?

**Comment by bbyhuy** [comment]
Hey @kai11 , i just got around to being able to test this. around 350 VMs was about the most I could get out of it before I started getting the error again. At some point we'd want to scale up to 1000 in parallel if possible.

edit:
To clarify, it seems running one-liner ansible commands works ok up to even 1000 in parallel. But, it is failing when running ansible-playbook. It gets through setup, then the very first task in the playbook will get the descriptor error.

**Comment by kai11** [comment]
@bbyhuy no idea why one-liner work but playbook is not. I'm not that familiar with Ansible.
My playbook to setup limits https://gist.github.com/kai11/bbbe7f86c05d278c17c1 also sets fs.file-max. Try that one?
@bcoca Can you comment?

**Comment by bcoca** [comment]
adhoc does not do a lot of work playbook does, it then requires less resources, persistence, etc so it is normal for it to work 'faster' and not have these kind of issues.

As for the fix, it would be nice to use kqueue/epoll when present and fallback to select otherwise (kqueue is also on the BSDs).

**Comment by kai11** [comment]
@bcoca https://github.com/kai11/ansible/blob/fix/select_fd_out_of_range_wip/lib/ansible/plugins/connection/ssh.py includes all 3 - KQueue, Poll (not epoll though) and Select.
Only Poll and Select were tested, KQueue is a guesswork - I did not tested it on FreeBSD.

**Comment by bcoca** [comment]
I would move the poller classes out of the ssh plugin, they might be useful for more than just that one.

**Comment by bcoca** [comment]
also i would only use epoll once the number of connections is very high (~ +1000)

**Comment by kamsz** [comment]
any update on that? i've the same problem with 600~ vms

**Comment by jhutar** [comment]
Same issue here with 300 clients (forks count set to 1000) - when I have set ulimit -n, -u and  sysctl fs.file-max to 65k and used https://github.com/ansible/ansible/issues/14143#issuecomment-183362205 , it works (have not tried without ulimit/sysctl not re-configured). Running on Red Hat Enterprise Linux 7 with `ansible-2.0.1.0-2.el7.noarch`.

**Comment by jhutar** [comment]
Hmm, well, with the change, after a while ansible playbook is consuming >1M open files. Not sure if that is expected:

```
# lsof | cut -d ' ' -f 1 | sort | uniq -c | sort -n | tail
   3855 nova-cond
   6650 qemu-kvm
   7069 neutron-s
   7072 memcached
  11577 nova-api
  12663 beam.smp
  20420 httpd
  26100 ssh
  54131 mysqld
1031994 ansible-p
```

and it is hanged too long on one of the tasks: https://github.com/redhat-performance/satperf/blob/master/scripts/clients.yaml#L59-L61 . Will probably kill it and workaround it somehow.

**Comment by mideeks** [comment]
Any update on this ticket? We're still stuck on 1.9.x because Ansible 2.x does not support more than a couple hundred hosts.

**Comment by lfbear** [comment]
thanks @kai11  I got it by your way

**Comment by discostur** [comment]
Any news on this? I'm also receiving this error ...

**Update:** Fixed it with @kai11 version, but i'm getting an "deprecation" warning now:

```
2016-12-02 14:51:55,796 p=34559 u=root |  [DEPRECATION WARNING]: ansible.utils.unicode.to_bytes is deprecated.  Use 
ansible.module_utils._text.to_bytes instead.
This feature will be removed in 
version 2.4.
```

So with version 2.4 this wouldn't work either.

**Comment by mideeks** [comment]
The changes from @kai11 does not include any calls to `to_bytes`. This must be a problem in the existing code in ansible devel. Diff of the changes: https://github.com/ansible/ansible/compare/devel...kai11:fix/select_fd_out_of_range_wip

Ansible 2.4 is not out or even the current devel version so it shouldn't be a problem for now.

This issue, among others, has unfortunately forced us to reconsider ansible. We will be moving off of it in the near future. I suggest others with large infra do the same.

**Comment by thomasbrezinski** [comment]
This has been open for almost a year and is a blocker for us upgrading to ansible 2. Any information about whether it's going to be addressed anytime soon?

**Comment by abadger** [comment]
/me crosses fingers

Do not get your hopes up too far but I've put together a PR that may fix this: https://github.com/ansible/ansible/issues/20934  It uses a backport of the selectors module from python-3.5 but works on the same principle as @kai11's code.

This is highly untested; I just whipped it together today when I noticed that there was a small, single file  backport of the selectors module on pypi that works with python-2.6 which we could bundle to prevent an increase in deps.  So far it has worked for me when testing smaller workloads.  If some of you with larger infrastructures that are hitting the file descriptor limits with select would care to test the new PR that would be an immense help.  The code is simple enough that, in theory, bcoca and I think this could make it into 2.3.0.  However, in practice, putting a change into the ssh connection plugin this close to freeze makes us nervous so we'll likely move it to be an ansible 2.4 feature if we find that the new code could be the source of new bugs.

Thanks for your patience.

**Comment by abadger** [comment]
I've merged #20934 into devel for now as we haven't encountered any problems and it seems to be the best way to get wider testing on it.  If it proves stable it will ship with 2.3.0  If we encounter any bugs in the run up to the release that seem to be caused by it, we'll revert for 2.3.x and revisit for 2.4.x.

**Comment by gzyeah** [comment]
well, we also met this issue in ansible 1.9 .  However, there we find workaround way is using strace while executing ansible. like: strace -f ansible . While then ansible will go with lower speed and lower FDs, but not interrupt again.  Strange for the relationship.
